Feature/heart

ibcylon opened the pull request:
## Motivation ⍰

- 하트 탭 베이스 구조 추가
- Rx Utils 추가
- chat API, Router 추가
<br>

## Key Changes 🔑

- 색상 스펠링 틀린 거 수정했습니다!

<br>

## To Reviewers 🙏🏻

- PR 좀 더 잘게 쪼개야 겠네요

<br>

## Linked Issue 🔗

-

ibcylon committed:
Add UnSelected Color Asset

ibcylon committed:
Add: TabBarController

ibcylon committed:
Feat: SingUp과 Main 분기 by accessToken

ibcylon committed:
Add: WhiteStrokeButton and BXCardView to Common Component

ibcylon committed:
Style: default to Dark Mode

ibcylon committed:
Add: TableView Utils

ibcylon committed:
Style: Fix the spelling neutral color from 'netural'

ibcylon committed:
Feat: Chat base

ibcylon committed:
Add: Chat Router

ibcylon committed:
Add Rx Utils for View LIfe cycle control

ibcylon committed:
Add: Notification Button to Nav Bar

ibcylon committed:
Add: Base Heart Scene

Minny27 commented:
navigator 클래스로 Push pop관리하는 거는 구현 안 해봐서 몰랐는데 신기하네요!
Application 싱글톤이 코디네이터같은 역할을 하는 걸까요?
모든 화면 전환을 담당하는 건지 궁금합니다.

Minny27 commented on Falling/Sources/Feature/Common/Application.swift:
```diff
@@ -13,10 +13,65 @@ final class Application {
+    fallingNavigationController.tabBarItem.imageInsets = .init(top: 5, left: 0, bottom: -5, right: 0)
+
+    let heartNavigationController = UINavigationController()
+    let heartNavigator = HeartNavigator(controller: heartNavigationController)
```
처음에는 기본 navigation controller를 파라미터로 전달하는게 어떤 의미로 전달하는지 몰랐는데
navigation stack이 쌓여있는 Navigation controller를 전달할 수도 있다고 생각하니까 이해가 되는군요.

Minny27 committed:
[#18]Refactor: Application 클래스 코드 함수화

- 탭 바 아이템 만드는 함수 구현
- 메인 및 마이 페이지 vc, viewmodel, navigator 생성

Minny27 committed:
[#18]Refactor: NavigationBarAppearance 설정 반복 코드 개선

- NavigationBarAppearance 모든 vc에서 동일하게 적용되야 할 거 같았음
- 그래서 슈퍼 클래스인 TFBaseViewController의 navigationSetting에 코드는 포함시킴

Minny27 commented:
TabBar Item 만드는 코드가 중복되서 createTabBarItem 메서드를 만들어서 개선했습니다.
```swift
//  Application.swift

final class Application {
  static let shared = Application()
  
  { ... }
  
  private func makeTabBarController() -> UIViewController {
    let mainNavigationController = UINavigationController()
    let heartNavigationController = UINavigationController()
    let chatNavigationController = UINavigationController()
    let myPageNavigationController = UINavigationController()
    
    let mainNavigator = MainNavigator(controller: mainNavigationController)
    let heartNavigator = HeartNavigator(controller: heartNavigationController)
    let chatNavigator = ChatNavigator(controller: chatNavigationController)
    let myPageNavigator = MyPageNavigator(controller: myPageNavigationController)
    
    let dependency = TabBarComponent()
    let tabBarController = TFTabBarController(dependency: dependency)
    
    tabBarController.viewControllers = [
      createTabBarItem(navigationController: mainNavigationController,
                       title: "폴링",
                       image: FallingAsset.Image.falling.image.withRenderingMode(.alwaysOriginal),
                       selectedImage: FallingAsset.Image.fallingSelected.image.withRenderingMode(.alwaysOriginal)),
      createTabBarItem(navigationController: heartNavigationController,
                       title: "하트",
                       image: FallingAsset.Image.heart.image.withRenderingMode(.alwaysOriginal),
                       selectedImage: FallingAsset.Image.heartSelected.image.withRenderingMode(.alwaysOriginal)),
      createTabBarItem(navigationController: chatNavigationController,
                       title: "채팅",
                       image: FallingAsset.Image.chat.image.withRenderingMode(.alwaysOriginal),
                       selectedImage: FallingAsset.Image.chatSelected.image.withRenderingMode(.alwaysOriginal)),
      createTabBarItem(navigationController: myPageNavigationController,
                       title: "마이",
                       image: FallingAsset.Image.more.image.withRenderingMode(.alwaysOriginal),
                       selectedImage: FallingAsset.Image.moreSelected.image.withRenderingMode(.alwaysOriginal))
    ]
    
    mainNavigator.toList()
    heartNavigator.toList()
    chatNavigator.toList()
    myPageNavigator.toList()
    return tabBarController
  }
  
  private func createTabBarItem(navigationController: UINavigationController,
                                title: String,
                                image: UIImage,
                                selectedImage: UIImage) -> UINavigationController {
    navigationController.tabBarItem = UITabBarItem(
      title: title,
      image: image,
      selectedImage: selectedImage
    )
    navigationController.tabBarItem.imageInsets = .init(top: 5, left: 0, bottom: -5, right: 0)
    return navigationController
  }
}
```

Minny27 commented:
<img src="https://github.com/THT-Team/THT-iOS/assets/68800789/6115aef1-80c1-4889-bbb4-0468a8c1f6a3" width=60%>

UINavigationBarAppearance 코드가 vc마다 중복적으로 나와서 모든 vc에서 동일하게 적용되야 한다고 생각되서
슈퍼클래스(TFBaseViewController)의 navigationSetting() 메서드에 해당 코드를 추가했습니다.

Minny27 commented on Falling/Sources/DataLayer/API/ChatAPI/ChatTarget.swift:
```diff
@@ -0,0 +1,53 @@
+  var task: Task {
+    switch self {
+    case .history(let request):
+      return .requestParameters(parameters: request.toDictionary(),
```
codable을 todictionary로 변환하는 방법이 있군요.. 배워갑니다

Minny27 commented on Project.swift:
```diff
@@ -73,7 +73,7 @@ final class BaseProjectFactory: ProjectFactory {
       "Item 4": "Pretendard-ExtraBold.otf"
     ],
-    "UIUserInterfaceStyle": "Light"
+    "UIUserInterfaceStyle": "Dark"
```
저희 다크모드만 지원하나요?

ibcylon replied:
그건 아닌데 현재 검정배경화면에서 statusBar가 안보여서 다크모드로했습니다.

Minny27 reviewed (approved):
코드 보고 많이 배워가네요
근데 파일이 너무 많아서 일일이 다 보지는 못했던 거 같아요ㅠ

ibcylon commented:
> 코드 보고 많이 배워가네요
> 근데 파일이 너무 많아서 일일이 다 보지는 못했던 거 같아요ㅠ

PR크기 1 commit은 너무 작은 것 같고
지금 관련없는 커밋까지 같이 PR 한 것 같아서
PR 당
300 - 500자안으로 해보겠습니다!

ibcylon commented:
> <img alt="" width="60%" src="https://user-images.githubusercontent.com/68800789/260521015-6115aef1-80c1-4889-bbb4-0468a8c1f6a3.png">
> UINavigationBarAppearance 코드가 vc마다 중복적으로 나와서 모든 vc에서 동일하게 적용되야 한다고 생각되서 슈퍼클래스(TFBaseViewController)의 navigationSetting() 메서드에 해당 코드를 추가했습니다.

네 감사합니다.
메인화면에는 아이콘 + 타이틀이라서 따로 분기가 필요해보여요!

ibcylon commented:
> navigator 클래스로 Push pop관리하는 거는 구현 안 해봐서 몰랐는데 신기하네요!
> Application 싱글톤이 코디네이터같은 역할을 하는 걸까요?
> 모든 화면 전환을 담당하는 건지 궁금합니다.

Application을 싱글톤으로 유지할 것인지 아니면 루트에서 주입받을 건지는 차후 고민이 필요해보이구요. 회원가입과 메인화면 분기만을 책임집니다.
나머지 화면전환은 하위 탭 네비게이터가 담당합니다.
또 push를 통한 화면 이동도 담당하게 될 수도 있습니다. 이것도 고민이 필요합니다.

Minny27 commented:
> > <img alt="" width="60%" src="https://user-images.githubusercontent.com/68800789/260521015-6115aef1-80c1-4889-bbb4-0468a8c1f6a3.png">
> > UINavigationBarAppearance 코드가 vc마다 중복적으로 나와서 모든 vc에서 동일하게 적용되야 한다고 생각되서 슈퍼클래스(TFBaseViewController)의 navigationSetting() 메서드에 해당 코드를 추가했습니다.
> 
> 네 감사합니다. 메인화면에는 아이콘 + 타이틀이라서 따로 분기가 필요해보여요!

타이틀, 아이콘은 어떻게 처리해야할 지 생각해봐야할 거 같아요!
